[Memory Snapshot] Make recordAnnotations callback initialize lazily

aaronenyeshi opened the pull request:
Summary: Make the recordAnnotations' Record function callback lazily initialize when record memory history starts. This will help reduce the impact on Time To First Batch metric.

Test Plan: CI and ran locally.

Differential Revision: D58875576

Pulled By: aaronenyeshi

aaronenyeshi commented:
We can ignore this test, since it seems to be very flaky: 
- [pull / linux-jammy-py3.8-gcc11 / test (distributed, 2, 2, linux.2xlarge)](https://hud.pytorch.org/pr/pytorch/pytorch/129242#26531387358) ([gh](https://github.com/pytorch/pytorch/actions/runs/9617476531/job/26531387358))
Process completed with exit code 1.

zdevito commented on torch/csrc/cuda/memory_snapshot.cpp:
```diff
@@ -117,6 +147,8 @@ void _record_memory_history(
     when = c10::cuda::CUDACachingAllocator::RecordContext::STATE;
   }
   at::globalContext().lazyInitCUDA();
+  c10::call_once(
```
I think this can all go in _initRecordAnnotations() by just putting the body inside a lambda?

aaronenyeshi replied:
Sounds good, updated.

aaronenyeshi committed:
[Memory Snapshot] Make recordAnnotations callback initialize lazily (#129242)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/129242

Make the recordAnnotations' Record function callback lazily initialize when record memory history starts. This will help reduce the impact on Time To First Batch metric.

Test Plan: CI and ran locally.

Reviewed By: davidberard98

Differential Revision: D58875576

Pulled By: aaronenyeshi
